File: 알고리즘/새로운시작/피자굽기.py
# https://www.acmicpc.net/problem/20061
import sys
from collections import deque
from bisect import bisect_left



input = sys.stdin.readline
D, N = map(int, input().split()) # 오븐의 깊이, 피자의 개수

oven = list(map(int, input().split())) # 오븐의 길이를 맞추기 위해 Dummy 값 넣기
pizza = deque(map(int, input().split()))

for i in range(1, len(oven)):
    if oven[i] > oven[i-1]:
        oven[i] = oven[i-1]

print(oven)
start = 0
end = D-1

while start <= end:
    mid = (start + end) // 2
    
    if 5 < oven[mid]:
        start = mid + 1
    else:
        end = mid
        
        
    print(mid)

File: Algorithm/C++/백준/구현/피자 굽기.cpp
// https://www.acmicpc.net/problem/1756
#include <iostream>
#include <vector>
using namespace std;
int D, N;
vector<int> oven;
int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    cin >> D >> N;
    oven.resize(D, 0);
    cin >> oven[0];

    // 오븐은 항상 오름차순으로 정렬 되어있다.
    for(int i = 1; i < D; i ++) {
        cin >> oven[i];
        // 오븐의 크기를 최소로 갱신한다.
        if(oven[i] > oven[i-1]) oven[i] = oven[i-1];
    }

    int depth = D;
    int pizza;
    for(int i = 0; i < N; i ++) {
        cin >> pizza;
        int start = 0;
        int end = depth - 1;

        // oven의 크기가 더 작은 위치를 찾는다.
        while(end >= start) {
            int mid = (start + end) / 2;
            if (pizza > oven[mid]) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        depth = end;
        if(depth < 0) {
            cout << "0";
            return 0;
        }
    }
    cout << depth + 1;
    return 0;
}